### PR TITLE
Add MAX_SLOPE capability to TEMPERATURE_WAIT

### PR DESCRIPTION
**Pull request by garethky**
TEMPERATURE_WAIT updated to support an optional MAX_SLOPE argument that allows a wait for the rate of change of temp to drop below that value.

To support this, both Heaters and PrinterSensorGeneric need a common way to provide the temperature slope to TEMPERATURE_WAIT. The code to do this was already in various places in Heaters.py. A new TemperatureSensorAdapter consolidates this functionality. This is used in both Heater and PrinterSensorGeneric so they now all have a known temperature slope which TEMPERATURE_WAIT can access via the adapter. This removes corresponding functionality from Heater and the control algorithms

Signed-off-by: Gareth Farrington <gareth@waves.ky>

This is an update for https://github.com/Klipper3d/klipper/pull/4273. I'm sorry I didn't have time to make the requested updates until now. Hopefully this version is closer to what you had in mind. It includes all of the changes you requested and tries to stay away from any unnecessary edits or changes to how Heaters or temperature sensors work now.

### CODE REVIEW

**Comment by garethky** [comment]
I finally had time to upgrade my printer and perform some testing. This update is tested to work on my printer. The smoothed slope seems to be less smooth that I remember before the upgrade so I'm going to assume that the default `smooth_time` has been shortened? Setting it to 2.0 on the sensor I'm testing with gives good performance.

* Fixed one outright typo which was throwing exceptions. 
* Fixed the initialization temp for all sensors to be `AMBIENT_TEMP` to not change the existing functionality.
* Changed the order of printing the `TEMPERATURE_WAIT` message so a message that shows a slope (or min/max temp) in the desired range prints before the function quits.

File: klippy/extras/heaters.py [hunk 1]
```diff
@@ -5,16 +5,107 @@
...
+        return {
+            'temperature': self.smoothed_temp,
+            'last_temperature': self.last_temperature,
+            'temperature_slope': self.temp_slope,
+        }
+
+######################################################################
+# Heater
+######################################################################
+
```
Comment (garethky) [inline klippy/extras/heaters.py:108]: extra blank line should be removed

**Comment by garethky** [comment]
I wasn't running `M190` in my printer and it turns out that was broken because one reference to `smoothed_temp` that wasn't converted. Fixed.

File: klippy/extras/heaters.py (hunk 1 above)
Comment (garethky) [reply klippy/extras/heaters.py:108]: Actually, the style seems to be to have a blank line. The merge just looks more complicated than it should be.

**Comment by garethky** [comment]
Withdrawn in favor of a macro:, see: https://github.com/Klipper3d/klipper/pull/4273#issuecomment-1248372083

**Comment by madorax251** [comment]
[gcode_macro HEAT_SOAK]
description: heats the bed for a while

variable_target_temp: 0
variable_stage: None ## heating -> soaking -> done -> None

## in seconds
variable_check_interval: 10
variable_soak_time_remaining: 0
variable_total_time_elapsed: 0

gcode:
    {% set TARGET = params.TARGET | default(0) | float %}
    {% set DURATION = (params.DURATION | default(5) | int) * 60 %} ## minutes to seconds

    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=target_temp         VALUE={ TARGET }
    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=stage               VALUE="'heating'"
    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=soak_time_remaining VALUE={ DURATION }
    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=total_time_elapsed  VALUE=0

    ;; fire up the heater
    SET_HEATER_TEMPERATURE HEATER=heater_bed TARGET={ TARGET }

    ;; run the fan to circulate air
    _SET_FAN_SPEED PERCENT=50

    ;; put the bed and nozzle where they're a safe distance apart
    G28
    CENTER

    M84 ;; turn off steppers

    UPDATE_DELAYED_GCODE ID=heat_soaker DURATION={ check_interval }

[gcode_macro CANCEL_HEAT_SOAK]
description: cancels an in-progress HEAT_SOAK cycle
gcode:
    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=stage VALUE="'cancel'"
    UPDATE_DELAYED_GCODE ID=heat_soaker DURATION=1

[delayed_gcode heat_soaker]
; ## debug
; { action_respond_info( printer['gcode_macro HEAT_SOAK'] | tojson )}
gcode:
    {% set heat_soak = printer['gcode_macro HEAT_SOAK'] %}

    ## update total time elapsed
    {% set total_time_elapsed = heat_soak.total_time_elapsed + heat_soak.check_interval %}
    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=total_time_elapsed VALUE={ total_time_elapsed }

    {% set stage = heat_soak.stage %}
    {% if stage == "heating" and printer.heater_bed.temperature >= heat_soak.target_temp %}
        {% set stage = "soaking" %}
    {% endif %}

    {% if stage == "soaking" %}
        ## update soak countdown
        {% set soak_time_remaining = [heat_soak.soak_time_remaining - heat_soak.check_interval, 0] | max %}
        SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=soak_time_remaining VALUE={ soak_time_remaining }

        {% if soak_time_remaining == 0 %}
            {% set stage = "done" %}
        {% endif %}
    {% endif %}

    SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=stage VALUE="'{ stage }'"

    {% if stage in ("done", "cancel") %}

        {% if stage == "cancel" %}
            {% set stage = "done" %}
            TURN_OFF_HEATERS
            M107 ; turn off fan

            M117 { "soak cancelled after ~%.1fm" | format(total_time_elapsed / 60.0) }
        {% else %}
            M117 { "soak complete after %.1fm" | format(total_time_elapsed / 60.0) }
        {% endif %}

        ## reset all state vars, except stage, which may be queried via the api
        SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=target_temp         VALUE=0
        SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=soak_time_remaining VALUE=0
        SET_GCODE_VARIABLE MACRO=HEAT_SOAK VARIABLE=total_time_elapsed  VALUE=0

    {% else %}

        {% if total_time_elapsed % 90 == 0 %}
            ## output status periodically
            {% if stage == "heating" %}
                M117 { "heating -- %.1fm elapsed" | format(total_time_elapsed / 60.0) }
            {% elif stage == "soaking" %}
                M117 { "soaking -- %.1fm remaining" | format(soak_time_remaining / 60.0) }
            {% endif %}
        {% endif %}

        ## trigger ourselves again
        UPDATE_DELAYED_GCODE ID=heat_soaker DURATION={ heat_soak.check_interval }

        ## dwell for 1ms to prevent from going idle
        G4 P1

    {% endif %}

**Pull request closed without merging**